x-pack/plugins/security_solution/scripts/run_cypress/parallel_serverless.ts

```diff
@@ -0,0 +1,531 @@
+      ? parseInt(process.env.BUILDKITE_PARALLEL_JOB_COUNT, 10)
+      : 1;
+    // An index which uniquely identifies this instance of the job
+    const chunkIndex: number = process.env.BUILDKITE_PARALLEL_JOB
+      ? parseInt(process.env.BUILDKITE_PARALLEL_JOB, 10)
+      : 0;
+
+    const nonSkippedSpecsForChunk: string[] = [];
+
+    for (let i = chunkIndex; i < nonSkippedSpecs.length; i += chunksTotal) {
+      nonSkippedSpecsForChunk.push(nonSkippedSpecs[i]);
+    }
+
+    return nonSkippedSpecsForChunk;
+  }
+};
+
+const encode = (str: string): string => Buffer.from(str, 'binary').toString('base64');
+
+const getApiKeyFromElasticCloudJsonFile = () => {
+  const userHomeDir = os.homedir();
+  try {
+    const jsonString = fs.readFileSync(path.join(userHomeDir, '.elastic/cloud.json'), 'utf-8');
+    const jsonData = JSON.parse(jsonString);
+    return jsonData.api_key.qa;
+  } catch (e) {
+    log.info('API KEY could not be found in ');
+    return null;
+  }
+};
+
+// Poller function that is polling every 20s, forever until function is resolved.
+async function poll<T>(
+  fn: () => Promise<T>,
+  retries: number = 200,
+  interval: number = 20000
+): Promise<T> {
+  return Promise.resolve()
+    .then(fn)
+    .catch(async function retry(err: Error): Promise<T> {
+      let remainingRetries = retries;
+      if (remainingRetries-- > 0) {
+        remainingRetries -= 1;
+        return new Promise((resolve) => setTimeout(resolve, interval)).then(fn).catch(retry);
+      }
+      throw err;
+    });
+}
+
+// Method to invoke the create environment API for serverless.
+async function createEnvironment(
+  baseUrl: string,
+  projectName: string,
+  runnerId: string,
+  apiKey: string,
+  ftrConfig: SecuritySolutionDescribeBlockFtrConfig,
+  onEarlyExit: (msg: string) => void
+): Promise<Environment> {
+  log.info(`${runnerId}: Creating environment ${projectName}...`);
+  const environment = {} as Environment;
+  const body = {
+    name: projectName,
+    region_id: DEFAULT_REGION,
+  } as CreateEnvironmentRequestBody;
+
+  const productTypes: ProductType[] = [];
+  ftrConfig?.productTypes?.forEach((t) => {
+    productTypes.push(t as ProductType);
+  });
+  if (productTypes.length > 0) body.product_types = productTypes;
+
+  await axios
+    .post(`${baseUrl}/api/v1/serverless/projects/security`, body, {
+      headers: {
+        Authorization: `ApiKey ${apiKey}`,
+        'Content-Type': 'application/json',
+      },
+    })
+    .then((response) => {
+      environment.name = response.data.name;
+      environment.id = response.data.id;
+      environment.region = response.data.region_id;
+      environment.es_url = `${response.data.endpoints.elasticsearch}:443`;
+      environment.kb_url = `${response.data.endpoints.kibana}:443`;
+      environment.product = response.data.type;
+    })
+    .catch((error) => {
+      onEarlyExit(`${error.code}:${error.data}`);
+    });
+  return environment;
```

Comment:
```suggestion
try {
  const response = await axios.post<SomeExpectedReturnType>(
    `${baseUrl}/api/v1/serverless/projects/security`,
    body,
    {
      headers: {
        Authorization: `ApiKey ${apiKey}`,
      },
    }
  );

  return {
    id: response.data.id,
    name: response.data.name,
    region: response.data.region_id,
    es_url: `${response.data.endpoints.elasticsearch}:443`,
    kb_url: `${response.data.endpoints.kibana}:443`,
    product: response.data.type,
  };
} catch (e) {
  throw new Error(`${e.code}:${e.data}`);
}
```

or try/catch can be omitted in favor of external error message handling. For example in a utility function.

Reply:
With try catch instead of an awat.then.catch the error that will be thrown in the exception is something of type:
response object is not parsable or name cannot be found in response object. 

Personally I would prefer to know WHY the error was thrown and this will take place if I know what error code and data I got from the rest api call I did to create the environment. 

As long as it comes to these, unless I am wrong, I would prefer keeping it this way.

Reply:
Fixed in commit : https://github.com/elastic/kibana/pull/169311/commits/d7613c83c18377ed29ec117643d912a047215e32